include/pybind11/detail/value_and_holder.h

```diff
@@ -74,5 +74,17 @@ struct value_and_holder {
     }
 };
 
+// This is a semi-public API to check if the corresponding instance has been constructed with a
+// holder. That is, if the instance has been constructed with a holder, the `__init__` method is
+// called and the C++ object is valid. Otherwise, the C++ object might only be allocated, but not
+// initialized. This will lead to **SEGMENTATION FAULTS** if the C++ object is used in any way.
+// Example usage: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#custom-type-setup
+//                for `tp_traverse` and `tp_clear` implementations.
+// WARNING: The caller is responsible for ensuring that the `reinterpret_cast` is valid.
+extern "C" [[nodiscard]] PYBIND11_NOINLINE bool is_holder_constructed(PyObject *obj) {
```

Comment:
Definitely remove the `extern "C"`, I don't think we need or want to call this function from C.

I'd also replace the `PYBIND11_NOINLINE` with `inline` for simplicity, that'll leave it to the compiler to decide how to handle this. (I've not seen real-life evidence that manual intervention actually helps.)